---
title: "Bioconductor Overview: edX 525.5x Part I"
author: "Vincent J. Carey"
package: edx2017.5x
output: 
  BiocStyle::html_document2:
    toc: true
    toc_float: true
abstract: >
  This document introduces concepts to be covered in the 5x module,
  *Introduction to Bioconductor*.  This overview document covers six
  key concepts: R language and packages, the central dogma of molecular
  biology, genome-scale assays, data structures for such assays,
  exploratory analysis, and hypothesis testing.  Our objective with
  this document is to
  ensure acquaintance with the higher-level concepts of the
  submodules to follow, with very light sketches of code in
  R that can be used to get some hands-on exposure to these
  concepts.  Some of the technical material on package
  structure and continuous integration can be safely skipped.
vignette: >
  %\VignetteIndexEntry{Bioc overview Part I}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---
```{r setup,echo=FALSE,results="hide"}
suppressWarnings({
suppressPackageStartupMessages({
library(png)
library(grid)
library(Homo.sapiens)
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg19)
library(RNAseqData.HNRNPC.bam.chr14)
library(GenomicAlignments)
library(Rsamtools)
library(ggbio)
library(erma)
library(VariantAnnotation)
#library(pRoloc)
suppressMessages({
ermaset=makeErmaSet()
})
})
})
```

# The R language and its packages and repositories

This course assumes a good working knowledge of the R language.
The Rstudio environment is recommended.  If you want to
work through a tutorial before proceeding, 
[Try R](http://tryr.codeschool.com/) is very comprehensive.

## Why R?

Bioconductor is based on [R](www.r-project.org).  Three key reasons for this are:

- R is used by many statisticians and biostatisticians to create algorithms that advance our ability to understand complex experimental data.
- R is highly interoperable, and fosters reuse of software components written in other languages. 
- R is portable to the key operating systems running on commodity computing equipment (Linux, MacOSX, Windows) and can be used immediately by beginners with access to any of these platforms.

In summary, R's ease-of-use and central role in statistics and "data science" make it a natural choice for a tool-set for use by biologists and statisticians confronting genome-scale experimental data.  Since the Bioconductor project's inception in 2001, it has kept pace with growing volumes
and complexity of data emerging in genome-scale biology.

## Functional object-oriented programming

R combines functional and object-oriented programming paradigms.^[[Chambers 2014](https://projecteuclid.org/euclid.ss/1408368569)]

- In functional programming, notation and program activity mimic the
concept of function in mathematics.  For example
```
square = function(x) x^2
```
is valid R code that defines the symbol `square` as a function that
computes the second power of its input.  The body of the function
is the program code `x^2`, in which `x` is a "free variable".
Once `square` has been defined in this way, `square(3)` has
value `9`.  We say the `square` function has been evaluated on
argument `3`.  **In R, all computations proceed by evaluation of functions.**

- In object-oriented programming, a strong focus is placed upon
formalizing data structure, and defining methods that take
advantage of guarantees established through the formalism.  This
approach is quite natural but did not get well-established in
practical computer programming until the 1990s.  As an
advanced example with Bioconductor, we will consider an 
approach to defining an "object" representing on the genome
of *Homo sapiens*:
```{r lkho}
library(Homo.sapiens)
class(Homo.sapiens)
methods(class=class(Homo.sapiens))
```
We say that `Homo.sapiens` is an **instance** of the `OrganismDb` 
**class**.  Every instance of this class will respond meaningfully
to the methods
listed above.  Each method is implemented as an R function.
What the function does depends upon the class of its arguments.
Of special note at this juncture are the methods
`genes`, `exons`, `transcripts` which will yield information about
fundamental components of genomes.  
These methods will succeed for human and
for other model organisms such as *Mus musculus*, *S. cerevisiae*,
*C. elegans*, and others for which the Bioconductor project and its contributors have defined `OrganismDb` representations.

## R packages, modularity, continuous integration

This section can be skipped on a first reading.

### Package structure

We can perform object-oriented functional programming with R
by writing R code.  A basic approach is to create "scripts" that
define all the steps underlying processes of data import and
analysis.  When scripts are written in such a way that they
only define functions and data structures, it becomes possible to
*package* them for convenient distribution to other users
confronting similar data management and data analysis problems.

The R software [packaging protocol](https://cran.r-project.org/doc/manuals/r-release/R-exts.html) specifies how source code in R and other languages can be organized together with metadata and documentation to foster
convenient testing and redistribution.  For example, an early
version of the package defining this document had the folder 
layout given below:

```
├── DESCRIPTION  (text file with metadata on provenance, licensing)
├── NAMESPACE    (text file defining imports and exports)
├── R            (folder for R source code)
├── README.md    (optional for github face page)
├── data         (folder for exemplary data)
├── man          (folder for detailed documentation)
├── tests        (folder for formal software testing code)
└── vignettes    (folder for high-level documentation)
    ├── biocOv1.Rmd
    ├── biocOv1.html
    └── wikiCentralDogma.png
```

The packaging protocol document "Writing R Extensions" provides
full details.  The R command `R CMD build [foldername]` will operate on the
contents of a package folder to create an archive that can
be added to an R installation using `R CMD INSTALL [archivename]`.
The R studio system performs these tasks with GUI elements.

### Modularity and formal interdependence of packages

The packaging protocol helps us to isolate software that
performs a limited set of operations, and to
identify the **version** of a program collection
that is inherently changing over time.  There is
no objective way to determine whether
a set of operations is the right size for packaging.
Some very useful packages carry out only a small number of
tasks, while others have very broad scope.  What is important
is that the package concept permits **modularization** of
software.  This is important in two dimensions: scope and time.
Modularization of scope is important to allow parallel independent
development of software tools that address distinct problems.
Modularization in time is important to allow identification of
versions of software whose behavior is stable.

### Continuous integration: testing package correctness and interoperability

The figure below is a snapshot of the [build report](http://bioconductor.org/checkResults/3.6/bioc-LATEST/) for the development branch of Bioconductor.

```{r lkci,fig=TRUE,echo=FALSE,fig.wide=TRUE, fig.cap="Continuous integration report for Bioconductor devel branch.  All packages are checked every night on three major computing platforms."}
im = readPNG("buildRep.png")
grid.raster(im)
```

The six-column subtable in the upper half of the display
includes a column "Installed pkgs", with entry 1857 for
the linux platform.  This number varies between platforms
and is generally increasing over time for the devel branch.


## Putting it together

Bioconductor's core developer group works hard to develop
data structures that allow users to work conveniently with
genomes and genome-scale data.  Structures are devised to
support the main phases of experimentation in genome scale biology:

- Parse large-scale assay data as produced by microarray or sequencer flow-cell scanners.
- Preprocess the (relatively) raw data to support reliable statistical interpretation.
- Combine assay quantifications with sample-level data to test hypotheses about relationships between molecular processes and organism-level characteristics such as growth, disease state.

In this course we will review the objects and functions that
you can use to perform these and related tasks in your own
research.

# The central dogma of molecular biology restated

Very simply, the central dogma states that DNA makes RNA makes
protein.  This, combined with the idea that cellular activities
are primarily driven by protein chemistry, motivates our
agenda.  DNA sequences are blueprints for protein collections,
but the path from DNA to functional protein and beyond,
to organismic structure and function, is long and
incompletely understood.

More detail is exposed in the following diagram.

```{r lklk,fig=TRUE,echo=FALSE,fig.wide=TRUE, fig.cap="Wikipedia schematic.  Chromosomes are composed of coding and non-coding DNA sequence.  Coding sequence includes exons and introns.  Exons are spliced and transcribed to messenger RNA (mRNA), which is then processed and translated to active protein."}
im = readPNG("wikiCentralDogma.png")
grid.raster(im)
```

In the past 10 years, many facets of the central dogma have
started to come into focus.  We will learn how to work with
data on DNA sequence, mRNA, and other genomic entities to
test hypotheses on biological processes (often in
the form of human diseases).  In this way we will strengthen
mankind's understanding of human health and the mechanisms
by which medical practices work to preserve health, and restore it 
when it has been lost.

# A collection of genome-scale assays

## The complete DNA sequence for an organism's reference genome

The figure below is a snapshot of use of the UCSC genome
browser, providing nucleotide-level depiction of sequence
and annotation for an expressed part of the gene ORMDL3.

```{r lkbrow,fig=TRUE,echo=FALSE,fig.wide=TRUE, fig.cap="DNA sequence and annotation for gene ORMDL3. This is a gene on chromosome 17 suspected of playing a role in the development of asthma."}
im = readPNG("ucscORMDL3.png")
grid.raster(im)
```

The sequence of letters ACTG along the top of the display are the
reference base calls for the human genome.  We will use Bioconductor
object to extract these in R, to illustrate 
the object-oriented programming concept noted above.

### An object representing a reference build of the human genome

The `BSgenome.Hsapiens.UCSC.hg19` package provides a representation
of the reference genome created at NCBI in 2009.  Individual
chromosomes (or more generally, annotated contiguous sequences,
some of which have yet to be positioned on chromosomes) can
be referenced using the `$` operator.
```{r lkhg}
library(BSgenome.Hsapiens.UCSC.hg19)
Hsapiens
Hsapiens$chr1
```

### A 'range' object identifying a set of genomic coordinates

The `GenomicRanges` package defines a family of data structures
useful for representing intervals on sequences.  A `GRanges`
instance can represent the coding region of a gene by defining
the name of the chromosome, the start and end positions of
the gene region, the strand on which the
gene resides, and the reference genome for which gene location
was determined.  Here we define the extent of coding
region for ORMDL3.
```{r dog}
library(GenomicRanges)
myr = GRanges("chr17", 
    IRanges(38077296, 38083884), strand="-", genome="hg19")
myr
```

### A request for nucleotide sequence

The nucleotide sequence for any interval is acquired
with the `getSeq` in the `Biostrings` package.
```{r doget}
getSeq(Hsapiens, myr)
```

## SNPs: cataloging population-level variation in DNA sequence

### Basic concepts

"SNP" stands for "single nucleotide polymorphism".  This is a simple
type of genetic variation useful in genetic epidemiology.
A SNP is a location in the (reference) genome that is _polymorphic_:
there is a given base present at the SNP location for the majority
of individuals, but some minority of individuals have an alternate base present.
The nucleotide codes are also called _alleles_.  We focus on
_biallelic_ SNP, for which the possible genotypes can be
generically described as A/A, A/B, or B/B.

For a given biallelic SNP,
the nucleotide most frequently observed is called the common allele,
and the less frequently observed nucleotide is called the rare allele.
A genomic location is regarded as a SNP if some 
fraction of individuals (typically at least 1\% of a geographically
or ethnically defined population) harbor the rare allele.
"SNP genotyping" refers to the determination of DNA content of individuals at
SNP locations -- individuals can be homozygous for the rare allele, heterozygous (have the reference
allele on one strand, the alternate allele on the other) or homozygous reference.

When multiple SNP are considered, it is possible to introduce the concept of
_phase_ for individuals heterozygous at multiple locations.  With phased genotypes
it is possible to distinguish configurations of nucleotides that occupy
the same chromosome, and are thus derived together from a given
parental chromosome.

### Epidemiological utility

Note 
[the 'SNPs in publications' glyph](https://genome.ucsc.edu/cgi-bin/hgc?c=chr17&l=38082721&r=38082892&o=38082806&t=38082807&g=pubsMarkerSnp&i=rs12603332&db=hg19) 
in the genome browser snapshot for ORMDL3 above.  The glyph harbors a link to
a series of publications that have shown health effects of genetic variation in this location.
The titles of the publications are
_Association of genetic variants in chromosome 17q21 and adult-onset asthma in a Chinese Han population_ (2011),
and 
_Allele-Specific Chromatin Remodeling in the ZPBP2/GSDMB/ORMDL3 Locus Associated with the Risk of Asthma and Autoimmune Disease_ (2009).
These are reports from basic projects in genetic epidemiology, stemming from observations of
genetic variation in a region of chromosome 17 connected with lung and autoimmune diseases.

### Concrete example

The `r Biocpkg("VariantAnnotation")` package includes a number
of utilities for working with and interpreting SNP data.  The following
output illustrates some of the fundamental components we've mentioned.

```{r docod,echo=FALSE,results="hide",cache=TRUE}
library(VariantAnnotation)
suppressMessages({
example(predictCoding)
})
```{r lkcoooo}
#
# after library(VariantAnnotation); example(predictCoding)
#
head(coding1[,c("REF", "ALT", "REFCODON", "VARCODON")])
```
<!--
GRanges object with 6 ranges and 4 metadata columns:
                  seqnames               ranges strand |            REF
                     <Rle>            <IRanges>  <Rle> | <DNAStringSet>
      rs114335781    chr22 [50301422, 50301422]      - |              G
        rs8135963    chr22 [50301476, 50301476]      - |              T
  22:50301488_C/T    chr22 [50301488, 50301488]      - |              C
  22:50301494_G/A    chr22 [50301494, 50301494]      - |              G
  22:50301584_C/T    chr22 [50301584, 50301584]      - |              C
      rs114264124    chr22 [50302962, 50302962]      - |              C
                                 ALT       REFCODON       VARCODON
                  <DNAStringSetList> <DNAStringSet> <DNAStringSet>
      rs114335781                  A            ATC            ATT
        rs8135963                  C            GCA            GCG
  22:50301488_C/T                  T            CCG            CCA
  22:50301494_G/A                  A            CAC            CAT
  22:50301584_C/T                  T            CCG            CCA
      rs114264124                  T            CGG            CAG
  -------
  seqinfo: 1 sequence from hg19 genome; no seqlengths
-->
This can be regarded as a table with six rows, one per SNP.  Some
of the SNP have 'rs' identifiers to the left, indicating presence in the
(NCBI dbSNP)[https://www.ncbi.nlm.nih.gov/projects/SNP/] catalog. 
Others have '22:' identifiers giving position and nucleotide options
for variants that have not yet been catalogued in dbSNP, perhaps
because they are not observed with sufficient frequency in populations
studied to date.  The REF and ALT columns give the reference and alternate
allele codes, and the REFCODON and VARCODON columns give the triplets
in the reference and alternate genome that will be translated to amino
acids.



## mRNA abundance estimates: depicting transcriptional states

The abundance of 
messenger RNA (mRNA) molecules in cells can be quantified
in various ways.  In general it is not possible to
obtain exact counts, but relative abundances 
are compared between samples after extensive
statistical preprocessing.  The overall process of
assaying single or multiple cells for mRNA abundance
is sometimes referred to as transcript profiling.

High-throughput sequencing of mRNA
(RNA-seq) has become the most common method
of measurement for current studies, but a large
archive of mRNA measurements obtained using
DNA microarrays is available at NCBI Gene Expression
Omnibus (GEO) and is useful for retrospective research.
We will examine both modalities of transcriptome
measurement, but for this introduction we focus
on RNA-seq.

### HNRNPC: a gene that may protect us from defective mRNA splicing

A 2013 paper of [Zarnack and colleagues](http://dx.doi.org/10.1016/j.cell.2012.12.023)
studies the role of the heterogeneous nuclear ribonucleoproteins C1 and C2
(coded for by gene HNRNPC) in the prevention of inclusion of cryptic Alu elements
in spliced transcripts.  One part of the study involves knockdown of HNRNPC
in HeLa cells, followed by genome-wide observation of frequency of cryptic Alu element
inclusion.  

The Bioconductor package `RNAseqData.HNRNPC.bam.chr14` includes an extract
from the RNA-seq data generated in this study.  There are 8 BAM files that have
been filtered and indexed to include information on mRNA molecules mapped
to the sequence of chromosome 14.  The protocol creates "paired-end" reads.
We can have a quick look at the short reads of mRNA sequence as follows:

```{r getlibs}
bf1 = RNAseqData.HNRNPC.bam.chr14_BAMFILES[1] # first of 8 files in package
r1 = readGappedReads(bf1)
r1[1:5]
qseq(r1[1:5])
```

We can demonstrate the detectability of mRNA coded for by HNRNPC
using the plotting package `ggbio`.  We use our knowledge of the location
of the gene to set up a scanning parameter to focus the data extract.
```{r domo,fig=TRUE,fig.height=4}
library(Rsamtools)
library(ggbio)
hnrnpc_param = 
   ScanBamParam(which=GRanges("chr14", IRanges(21.67e6,21.74e6)))
limr1 = readGappedReads(bf1, param=hnrnpc_param, use.names=TRUE)
autoplot(limr1, geom="line", stat="coverage") + ylim(0,900)
```

In exercises you will verify that the knockdown experiment succeeded.

<!--
pr1 = readGAlignmentPairs(bf1, param=ScanBamParam(what="seq"))
-->

## Epigenomics: how regulatory factors bind to DNA to control transcription and translation

Numerous genome-scale assays have been devised 
to investigate epigenetic phenomena.  The methods and
implications are diverse and we give two very
limited examples to sketch the landscape.

### Cell-type specific regulatory roles of chromosomal regions

How do cells that possess a common blueprint (DNA sequence) differentiate
into units performing distinct roles in different organs or
organ components (for example, eye and tooth, or cornea and lens)?
The linear DNA sequence shared by all cells in an individual cannot
explain this.  The [Epigenomics Roadmap project](http://www.roadmapepigenomics.org/)
collects results of thousands of experiments that help to characterize
aspects of DNA accessibility and modification that regulate gene
expression and cell function.  The `r Biocpkg("erma")` package
includes data and code to facilitate exploration of some of the
outcomes of the project.  In the figure below, the regulatory
architecture upstream of ORMDL3 is sketched for five different
cell types.  We use the `csProfile` function of `r Biocpkg("erma")`
to depict chromatin state profiles as computed using
the [chromImpute algorithm](http://www.biolchem.ucla.edu/labs/ernst/ChromImpute/)
of [Ernst and Kellis](http://www.nature.com/nbt/journal/v33/n4/full/nbt.3157.html).
Each horizontal band corresponds to a cell type, with colored regions
indicating the predicted functional role of epigenetically modified
chromosomal sequence, over genomic coordinates given at the bottom.

```{r doer,eval=FALSE}
library(erma)
ermaset = makeErmaSet()
csProfile(ermaset[,1:5], "ORMDL3", shortCellType=FALSE)
```
```{r lkprom1,fig=TRUE,results="hide",echo=FALSE}
suppressMessages({
csProfile(ermaset[,1:5], "ORMDL3", shortCellType=FALSE)
})
```
The black vertical line is positioned at the actual transcription
start site (TSS) for ORMDL3, shared for all cell types.  We see that T and
B cells may be distinguished by the presence or absence of a
"poised promoter" type region about 1500 bases upstream of the TSS.
Decoding of the
functional terminology of
the color key
can be done using the [erma vignette](http://bioconductor.org/packages/release/bioc/vignettes/erma/inst/doc/erma.html).

The use of blocked bands to illustrate variation in epigenomic
characteristics over genomic coordinates is extremely crude.  It
reflects the reduction of information starting with short reads
obtained
in ChIP-seq experiments, proceeding through read alignment and peak
scoring, culminating in machine learning of associations between
peak configurations of multiple epigenomic marks and functional
state of chromosomal sequence on which the marks are found.
Statistical characterization of uncertainty of the inferred
states requires additional research.  Additionally, understanding
of contexts of variability (between individuals, within
individuals over time) of chromatin states as inferred
in this display, will require additional work.

### Analysis of differential methylation with whole-genome bisulfite sequencing

The role of epigenetic modifications of genes involved in
cancer is a target of intensive study.  Kasper Hansen and colleagues
developed the `r Biocpkg("bsseq")` package to help identify
and interpret differential methylated regions of the genome,
with application to comparison of matched tumor-normal sample
pairs.  As we have seen before, the data are readily
oriented along genomic coordinates.  Statistical methods
for smoothing noisy measures of methylation frequency and
testing for tumor-normal disparity lead to displays like
the following.

Here we attach the packages that provide example data
and analytic functions for an experiment involving three
matched tumor-normal pairs, and examine the basic
sequencing data.
```{r lkbsseq,fig=TRUE}
library(bsseq)
library(bsseqData)
BS.cancer.ex
```
The display of `BS.cancer.ex` indicates its potential
for a role in a 'smoothing' process.  The sample
information can be obtained with the `colData`
method.
```{r lkcold}
colData(BS.cancer.ex)
```

In exercises we will examine these objects in more details.
The ultimate result of the workflow for whole genome
bisulfite sequencing is a catalog of differentially
methylated regions.  In the `bsseqData` package,
we have snapshots of data at various stages.  The .tstat
object includes results of large-scale testing for
differential methylation between tumor and
normal tissues, accounting for biological
variability between samples.  The .fit object includes
the result of sample-specific smoothing of methylation
measures, over genomic coordinates.

```{r dodmr,fig=TRUE}
dmrs0 <- dmrFinder(BS.cancer.ex.tstat)
dmrs <- subset(dmrs0, n >= 3 & abs(meanDiff) >= 0.1)
BS.cancer.ex.fit$col = rep(c("red", "blue"), c(3,3))
plotRegion(BS.cancer.ex.fit, dmrs[1,], 
      extend = 5000, addRegions = dmrs)
```

## Other crucial assay classes: Proteomics, metabolomics, microbiomics, whole-genome editing

### Organelle proteomics

Thorough characterization of gene function depends upon
many sorts of experiments at different scales.  The `r Biocpkg("pRoloc")`
package addresses an approach to localization of proteins to 
organelles.  An interesting resource provided here is a catalog
of organelle marker proteins:

```{r drop,echo=FALSE,results="hide"}
#detach("package:png", unload=TRUE)
detach("package:ggbio", unload=TRUE)
suppressWarnings({
suppressPackageStartupMessages({
library(pRoloc)
})
})
```

```{r domark}
library(pRoloc)
pm = pRolocmarkers("hsap") # H. sapiens
table(pm)
names(head(pm[pm=="Proteasome"]))
```

These are UNIPROT identifiers whose association to genes or GO categories
or pathways can be assessed with annotation infrastructure that we will
review later in the course.

```{r drop2,echo=FALSE,results="hide"}
detach("package:pRoloc", unload=TRUE)
suppressWarnings({
suppressPackageStartupMessages({
library(phyloseq)
})
})
```

### Metabolomics: small-molcule fingerprints of cellular processes

Mass spectrometry is typically employed to 
assay biosamples for metabolites.  A number of Bioconductor
packages address capture and interpretation of mass spec
experiments for proteomic or metabolomic experiments.
See the [metabolomic 
task view](http://bioconductor.org/packages/release/BiocViews.html#___Metabolomics) for examples and vignettes.  Some details
will be explored later in the course.

The [hmdbQuery package](https://github.com/vjcitn/hmdbQuery) (presently
available only in github, in review at Bioconductor) helps with downstream
interpretation of metabolomics experiments.  
The package is an interface to the [Human Metabolome Database](http://www.hmdb.ca).
Two basic
utilities are provided: a collection of tables mapping
HMDB-catalogued
metabolites to proteins, genes, and diseases, and a
function that will retrieve a detailed record about
a metabolite, given the HMDB identifier.
The tables provided are facets of a snapshot of the HMDB
taken 23 Sept. 2017.  These can be updated using scripts
provided in the package.

Let's begin with a peek at the metabolite-disease mapping
table.

```{r set2,echo=FALSE,results="hide"}
suppressMessages({
suppressPackageStartupMessages({
library(hmdbQuery)
data(hmdb1)
data(hmdb_disease)
library(rtracklayer)
library(CrispRVariants)
library(sangerseqR)
})
})
```
```{r lkdisss}
dim(hmdb_disease)
head(hmdb_disease)
```


```{r lkent,eval=FALSE}
library(hmdbQuery)
HmdbEntry(id="HMDB0000001")
```
```{r lkd,echo=FALSE}
hmdb1
```

### Microbiome analysis

A nice overview of history and concepts of
microbiome studies is given by
[Morgan and Huttenhower (2012)](doi:10.1371/journal.pcbi.1002808).
These authors describe how classical microbiology was highly
dependent on cell culturing methodologies.  Current approaches
to identifying
microbial communities use DNA sequencing to identify
organisms for which culturing protocols do not exist.
Reference genome sequences are available for only a small
minority of microbes present in, for example, the human
gut.  Thus methods of phylogenetics are applied to derive
inferences on existence and abundance of species in microbiome
samples.  A key step is forming clusters of sequences using
measures of sequence similarity, thus defining _operational
taxonomic units_ (OTUs).

An illustrative example is given in the `r Biocpkg("phyloseq")`
package.  A dataset is provided that was used to demonstrate
elevation in the abundance of _Fusobacterium_ in colon
tumor tissue when compared to matched controls.

In the following code snippet, we use `r Biocpkg("phyloseq")`
to acquire data generated using [QIIME](http://www.qiime.org).
```{r lkph}
library(phyloseq)
filepath = system.file("extdata", 
   "study_1457_split_library_seqs_and_mapping.zip", package="phyloseq")
kostic = microbio_me_qiime(filepath)
```
An excerpt from the taxonomy table associated with this experiment
shows that OTUs are linked to known organism classes with
varying levels of specificity.
```{r lktax}
tax_table(kostic)[1:4,]
```
The read counts for two different OTUs in 10 samples follow:
```{r lkct}
otu_table(kostic)["304309",1:10]
otu_table(kostic)["358030",1:10]
```



### A glimpse of CRISPR


```{r docr,cache=TRUE,echo=FALSE,results="hide"}
suppressWarnings({
suppressMessages({
suppressPackageStartupMessages({
library(CrispRVariants)
library(sangerseqR)

# List AB1 filenames, get sequence names,  make names for the fastq files
# Note that we only include one ab1 file with CrispRVariants because
# of space constraints.  All bam files are included

data_dir <- system.file(package="CrispRVariants", "extdata/ab1/ptena")
fq_dir <- tempdir()
ab1_fnames <- dir(data_dir, "ab1$", recursive=TRUE, full=TRUE)
sq_nms <- gsub(".ab1","",basename(ab1_fnames))

# Replace spaces and slashes in filename with underscores
fq_fnames  <- paste0(gsub("[\ |\\/]", "_", dirname(ab1_fnames)), ".fastq")

# abifToFastq to read AB1 files and write to FASTQ
dummy <- mapply( function(u,v,w) {
        abifToFastq(u,v,file.path(fq_dir,w))
}, sq_nms, ab1_fnames, fq_fnames)


## ---- message=FALSE--------------------------------------------------------
# The metadata and bam files for this experiment are included with CrispRVariants
md_fname <- system.file(package="CrispRVariants", "extdata/metadata/metadata.xls")
md <- gdata::read.xls(md_fname, 1)
md

# Get the bam filenames from the metadata table
bam_dir <- system.file(package="CrispRVariants", "extdata/bam")
bam_fnames <- file.path(bam_dir, md$bamfile)

# check that all files exist
all( file.exists(bam_fnames) )

## ---- message=FALSE--------------------------------------------------------
# Represent the guide as a GenomicRanges::GRanges object
gd_fname <- system.file(package="CrispRVariants", "extdata/bed/guide.bed")
gd <- rtracklayer::import(gd_fname)
gd

## ---- message=FALSE--------------------------------------------------------
gdl <- GenomicRanges::resize(gd, width(gd) + 10, fix = "center")

## ---- eval=FALSE-----------------------------------------------------------
#  system("samtools faidx GRCHz10.fa.gz")
#  
#  reference=system(sprintf("samtools faidx GRCHz10.fa.gz %s:%s-%s",
#                           seqnames(gdl)[1], start(gdl)[1], end(gdl)[1]),
#                   intern = TRUE)[[2]]
#  
#  # The guide is on the negative strand, so the reference needs to be reverse complemented
#  reference=Biostrings::reverseComplement(Biostrings::DNAString(reference))
#  save(reference, file = "ptena_GRCHz10_ref.rda")

## --------------------------------------------------------------------------
ref_fname <- system.file(package="CrispRVariants", "extdata/ptena_GRCHz10_ref.rda")
load(ref_fname)
reference

## ---- tidy = FALSE---------------------------------------------------------
# First read the alignments into R.  The alignments must include
# the read sequences and the MD tag
alns <- GenomicAlignments::readGAlignments(bam_fnames[[1]], 
          param = Rsamtools::ScanBamParam(tag = "MD", what = c("seq", "flag")),
          use.names = TRUE)

# Then reconstruct the reference for the target region.
# If no target region is given, this function will reconstruct
# the complete reference sequence for all reads.
rfa <- refFromAlns(alns, gdl)

# The reconstructed reference sequence is identical to the sequence
# extracted from the reference above
print(rfa == reference)

## ---- message=FALSE--------------------------------------------------------
# Note that the zero point (target.loc parameter) is 22
crispr_set <- readsToTarget(bam_fnames, target = gdl, reference = reference,
                            names = md$Short.name, target.loc = 22)
crispr_set

# The counts table can be accessed with the "variantCounts" function
vc <- variantCounts(crispr_set)
print(class(vc))

## ---- eval = FALSE---------------------------------------------------------
#  # In R
#  library(GenomicFeatures)
#  gtf_fname <- "Danio_rerio.GRCz10.81_chr17.gtf"
#  txdb <- GenomicFeatures::makeTxDbFromGFF(gtf_fname, format = "gtf")
#  saveDb(txdb, file= "GRCz10_81_chr17_txdb.sqlite")

## ---- echo=FALSE, message=FALSE--------------------------------------------
library(GenomicFeatures)
txdb_fname <- system.file("extdata/GRCz10_81_ptena_txdb.sqlite", 
                          package="CrispRVariants")
txdb <- loadDb(txdb_fname)

## ---- message = FALSE------------------------------------------------------
# The gridExtra package is required to specify the legend.key.height 
# as a "unit" object.  It is not needed to call plotVariants() with defaults
library(gridExtra)

# Match the clutch id to the column names of the variants
group <- md$Group

## ----ptena-plot, fig.width = 8.5, fig.height = 7.5, message = FALSE, fig.cap = "(Top) schematic of gene structure showing guide location (left) consensus sequences for variants (right) variant counts in each embryo."----
})
})
})
```
```{r fopl,fig=TRUE,echo=FALSE}
suppressWarnings({
suppressMessages({
p <- plotVariants(crispr_set, txdb = txdb, gene.text.size = 8, 
    row.ht.ratio = c(1,8), col.wdth.ratio = c(4,2),
    plotAlignments.args = list(line.weight = 0.5, ins.size = 2, 
                               legend.symbol.size = 4),
    plotFreqHeatmap.args = list(plot.text.size = 3, x.size = 8, group = group, 
                                legend.text.size = 8, 
                                legend.key.height = grid::unit(0.5, "lines"))) 
})
})
```

# Data structures for genome-scale assays

# Exploring genome-scale assay data

# Testing statistical hypotheses with genome-scale assays

